### ISSUE TITLE
n8n deploued on docker seems to add extra new line to username and password when using secrets

### ISSUE THREAD

**Issue by bartoszrapp** [n8n-io/n8n#10818]
### Bug Description

When you try to deploy n8n with Postgres and you want to use secrets to safely pass the DB credentials is seems the \n is added at the end of content read from /run/secrets/.... resulting in authorization error. With user and password passed explicite using env variable everything works as expected.

n8n_postgres  | 2024-09-14 16:52:27.159 UTC [68] FATAL:  password authentication failed for user "n8n_db_user
n8n_postgres  |         "
n8n_postgres  | 2024-09-14 16:52:27.159 UTC [68] DETAIL:  Role "n8n_db_user
n8n_postgres  |         " does not exist.
n8n_postgres  |         Connection matched file "/var/lib/postgresql/data/pg_hba.conf" line 128: "host all all all scram-sha-256"
n8n_app       | Error: There was an error initializing DB
n8n_app       | DatabaseError: password authentication failed for user "n8n_db_user
n8n_app       | "

### To Reproduce

services:
  n8n:
    image: docker.n8n.io/n8nio/n8n
    container_name: n8n_app
    restart: always
    ports:
      - 9090:5678
    environment:
      - N8N_PROXY_HOPS=1
      - N8N_SECURE_COOKIE=false
      - N8N_PORT=5678
      - N8N_PROTOCOL=https
      - DB_TYPE=postgresdb
      - DB_POSTGRESDB_PORT=5432
      - DB_POSTGRESDB_HOST=postgres
      - DB_POSTGRESDB_DATABASE=n8n
      - DB_POSTGRESDB_USER_FILE=/run/secrets/db_username
      - DB_POSTGRESDB_PASSWORD_FILE=/run/secrets/db_password
    volumes:
      - n8n_data:/home/node/.n8n
      - /home/docker/storage/n8n-local-files:/files
    networks:
      - proxy_bus_network
      - internal_n8n
    secrets:
      - db_username
      - db_password
    depends_on:
      postgres:
        condition: service_started

  postgres:
    image: postgres:16
    restart: always
    container_name: n8n_postgres
    environment:
     - POSTGRES_DB=n8n
     - POSTGRES_USER_FILE=/run/secrets/db_username
     - POSTGRES_PASSWORD_FILE=/run/secrets/db_password
    volumes:
      - n8n_postgres_data:/var/lib/postgresql/data
    networks:
      - internal_n8n
    secrets:
      - db_username
      - db_password

volumes:
  n8n_data:
    name: n8n_data
  n8n_postgres_data:
    name: n8n_postgres_data

networks:
  proxy_bus_network:
    name: proxy_bus_network
    external: true
  internal_n8n:
    name: internal_n8n

secrets:
  db_username:
    file: ./db_username.txt
    name: n8n_pg_username
  db_password:
    file: ./db_password.txt
    name: n8n_pg_password

### Expected behavior

Using DB_POSTGRESDB_USER_FILE and DB_POSTGRESDB_PASSWORD_FILE should work the same as using DB_POSTGRESDB_USER and DB_POSTGRESDB_PASSWORD

### Operating System

Ubuntu Linux 24.04 LTS

### n8n Version

1.58.2

### Node.js Version

20.15.0

### Database

PostgreSQL

### Execution mode

main (default)

**Comment by Joffcom** [comment]
Hey @bartoszrapp,

We have created an internal ticket to look into this which we will be tracking as "GHC-232"

**Comment by bartoszrapp** [comment]
As an additional information. It is for sure a problem on n8n side. Postgres set with secrets and _FILE and n8n set with regular password env variable works perfectly fine. Only when I use xxx_FILE on n8n the error manifestates.

**Comment by Joffcom** [comment]
Hey @bartoszrapp 

Just to confirm your secret file doesn't have a new line character at the end of it right? I will test this on Monday but I have seen some text editors automatically add a return when saving so want to quickly rule that out.

**Comment by bartoszrapp** [comment]
No there is no new line. I have checked it few times. I have even added a new line once to check the result in the logs. It was than with extra new line. Looked like this:
n8n_postgres | 2024-09-14 16:52:27.159 UTC [68] DETAIL: Role "n8n_db_user
n8n_postgres | 
n8n_postgres | " does not exist.

**Comment by Joffcom** [comment]
Hey @bartoszrapp, 

Perfect, I have taken a look and managed to reproduce the issue.

**Comment by bartoszrapp** [comment]
Hey @Joffcom,

Thank you for the update. I hope this will be an easy fix and nothing serious that would require a lot of work :)

**Comment by chmelevskij** [comment]
Seems that problem is still there.

Secrets seem to be created correctly, for my case they are created with ansible and there is trimming at that stage
```
   - name: Create required Docker secrets
      command: >
        docker secret create {{ item.name }} -
      args:
        stdin: '{{ item.value | trim }}'
      when: secret_check.results[item.index].rc != 0
      loop:
        - {
            name: 'postgres_user',
            value: '{{ postgres_user | trim }}',
            index: 0,
          }
        - {
            name: 'postgres_password',
            value: '{{ postgres_password | trim }}',
            index: 1,
          }
        - { name: 'postgres_db', value: '{{ postgres_db | trim }}', index: 2 }
        - {
            name: 'postgres_non_root_user',
            value: '{{ postgres_non_root_user | trim }}',
            index: 3,
          }
        - {
            name: 'postgres_non_root_password',
            value: '{{ postgres_non_root_password | trim }}',
            index: 4,
          }
        - {
            name: 'encryption_key',
            value: '{{ encryption_key | trim }}',
            index: 5,
          }
      register: create_secrets
```

Also connecting to postgres container and then using credentials works to connect to the database.

Similar if using env variables for n8n and secrets for everything else it starts just fine.

**Comment by jamesyc** [comment]
n8n setup does not work. 

I am following the instructions here: https://docs.n8n.io/hosting/installation/docker/

which tells us to use the docker-compose.yml and .env file from here: https://github.com/n8n-io/n8n-hosting/tree/main/docker-compose

However, trying to start the docker container for the first time fails. Running `docker logs n8n-postgres-1` gives me: 
    2025-04-28 11:28:41.773 UTC [41] FATAL:  password authentication failed for user "user"
    2025-04-28 11:28:41.773 UTC [41] DETAIL:  Role "user" does not exist.
	    Connection matched file "/var/lib/postgresql/data/pg_hba.conf" line 128: "host all all all scram-sha-256"

**Comment by vedangdate** [comment]
Any update on this?

**Comment by avendretter** [comment]
having the same issue. 
Running this setup: https://github.com/n8n-io/n8n-hosting/tree/main/docker-compose/withPostgres

I did not change the .env file on purpose. 

> ❯ dc up
[+] Building 0.0s (0/0)                                                                                                                                               docker:default
[+] Running 3/3
 ✔ Network withpostgres_n8n           Created                                                                                                                                   0.0s
 ✔ Container withpostgres-postgres-1  Created                                                                                                                                   0.1s
 ✔ Container withpostgres-n8n-1       Created                                                                                                                                   0.0s
Attaching to withpostgres-n8n-1, withpostgres-postgres-1
withpostgres-postgres-1  |
withpostgres-postgres-1  | PostgreSQL Database directory appears to contain a database; Skipping initialization
withpostgres-postgres-1  |
withpostgres-postgres-1  | 2025-06-09 15:51:21.523 UTC [1] LOG:  starting PostgreSQL 16.9 (Debian 16.9-1.pgdg120+1) on aarch64-unknown-linux-gnu, compiled by gcc (Debian 12.2.0-14) 12.2.0, 64-bit
withpostgres-postgres-1  | 2025-06-09 15:51:21.523 UTC [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
withpostgres-postgres-1  | 2025-06-09 15:51:21.523 UTC [1] LOG:  listening on IPv6 address "::", port 5432
withpostgres-postgres-1  | 2025-06-09 15:51:21.526 UTC [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
withpostgres-postgres-1  | 2025-06-09 15:51:21.530 UTC [29] LOG:  database system was shut down at 2025-06-09 15:50:41 UTC
withpostgres-postgres-1  | 2025-06-09 15:51:21.536 UTC [1] LOG:  database system is ready to accept connections
withpostgres-n8n-1       | No encryption key found - Auto-generating and saving to: /home/node/.n8n/config
withpostgres-n8n-1       | Permissions 0644 for n8n settings file /home/node/.n8n/config are too wide. This is ignored for now, but in the future n8n will attempt to change the permissions automatically. To automatically enforce correct permissions now set N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS=true (recommended), or turn this check off set N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS=false.
withpostgres-n8n-1       | Initializing n8n process
withpostgres-postgres-1  | 2025-06-09 15:51:29.607 UTC [41] FATAL:  password authentication failed for user "User"
withpostgres-postgres-1  | 2025-06-09 15:51:29.607 UTC [41] DETAIL:  Role "User" does not exist.
withpostgres-postgres-1  | 	Connection matched file "/var/lib/postgresql/data/pg_hba.conf" line 128: "host all all all scram-sha-256"
withpostgres-n8n-1       | There was an error initializing DB
withpostgres-n8n-1       | password authentication failed for user "User"
withpostgres-n8n-1 exited with code 1
withpostgres-n8n-1       | Permissions 0644 for n8n settings file /home/node/.n8n/config are too wide. This is ignored for now, but in the future n8n will attempt to change the permissions automatically. To automatically enforce correct permissions now set N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS=true (recommended), or turn this check off set N8N_ENFORCE_SETTINGS_FILE_PERMISSIONS=false.
withpostgres-n8n-1       | User settings loaded from: /home/node/.n8n/config
withpostgres-n8n-1       | Last session crashed
withpostgres-n8n-1       | Initializing n8n process
.....

**Comment by vedangdate** [comment]
solved it on my end
I deleted all the blank lines from my .env file and it started working. I think the issue was specifically a blank line after the password prop but to make it a convention for my project, I stopped adding blank lines.

**Comment by fredericopadrao** [comment]
I'm having the exact same issue over here, on n8n 1.99.1, just as @bartoszrapp. Postgres is reading the password without any problem from a secret file, but when I try to get the same secret on n8n from the same file to fill the env "DB_POSTGRESDB_PASSWORD_FILE:" it says that the password is incorrect.

**Comment by TreyDong** [comment]
It's solved.
docker-compose.yml should like this：

```
 n8n:
    image: n8nio/n8n:1.101.1
    restart: always
    environment:
      - DB_TYPE=postgresdb
      - DB_POSTGRESDB_HOST=postgres
      - DB_POSTGRESDB_PORT=5432
      - DB_POSTGRESDB_DATABASE=${POSTGRES_DB}
      - DB_POSTGRESDB_USER=${POSTGRES_USER}
      - DB_POSTGRESDB_PASSWORD=${POSTGRES_PASSWORD}
```

**Comment by gpongelli** [comment]
Hi, issue seems fixed.

I've a docker-compose file with
```yaml
  n8n:
    container_name: n8n
    image: n8nio/n8n:1.105.2
    restart: unless-stopped
    ports:
      - 5678:5678
    env_file:
      - ./services/n8n/n8n.env
    environment:
      DB_POSTGRESDB_DATABASE: /run/secrets/n8n-pg-db
      DB_POSTGRESDB_USER: /run/secrets/n8n-pg-user
      DB_POSTGRESDB_PASSWORD: /run/secrets/n8n-pg-pass
    volumes:
       <REDACTED>
    secrets:
      - n8n-pg-db
      - n8n-pg-user
      - n8n-pg-pass

....

secrets:
...
  n8n-pg-db:
    file: ./services/n8n/.env.pgdb
  n8n-pg-user:
    file: ./services/n8n/.env.pguser
  n8n-pg-pass:
    file: ./services/n8n/.env.pgpass
```

those three files `.env.pgpass` , `.env.pguser` and `.env.pgdb` contain ONLY the needed data, no newline at the end.

`n8n.env` file referenced in env_file does NOT contain the three variables.

the container does not start and its log contains repeatedly this:
```txt
Initializing n8n process
There was an error initializing DB
password authentication failed for user "/run/secrets/n8n-pg-user"
```
It seems n8n does not read the content, but uses the value as is.



Changing to 
```yaml
    environment:
      DB_POSTGRESDB_DATABASE_FILE: /run/secrets/n8n-pg-db
      DB_POSTGRESDB_USER_FILE: /run/secrets/n8n-pg-user
      DB_POSTGRESDB_PASSWORD_FILE: /run/secrets/n8n-pg-pass
```
fixes the issue.

**Comment by elcid2012** [comment]
> Hi, issue seems fixed.
> 
> [ ... ]
> 
>     environment:
>       DB_POSTGRESDB_DATABASE: /run/secrets/n8n-pg-db
>       DB_POSTGRESDB_USER: /run/secrets/n8n-pg-user
>       DB_POSTGRESDB_PASSWORD: /run/secrets/n8n-pg-pass
> 
> 
> Changing to
> 
>     environment:
>       DB_POSTGRESDB_DATABASE_FILE: /run/secrets/n8n-pg-db
>       DB_POSTGRESDB_USER_FILE: /run/secrets/n8n-pg-user
>       DB_POSTGRESDB_PASSWORD_FILE: /run/secrets/n8n-pg-pass
> fixes the issue.

No, that fixes your issue, because you didn't have your original environment variables defined properly for using _FILE and docker secrets.

Also this issue is still not fixed. I can get n8n to see the database by setting DB_POSTGRESDB_PASSWORD via an environment variable or stored directly in the docker-compose.yml file. As soon as I switch to DB_POSTGRESDB_PASSWORD_FILE with the same password, authentication errors.

**Comment by elcid2012** [comment]
We're all circling around the answer here, but using the clues I searched and found this: https://askubuntu.com/questions/13317/how-to-stop-gedit-gvim-vim-nano-from-adding-end-of-file-newline-char

Here's the actual fix:

`nano -L db_password`

Then insert your password into the file, and save as usual.

The `-L` flag forces nano to NOT insert new lines at the end of the file. You can verify this with this psuedo code:

```
user@host:~/$ nano db_password
# enter "1234" into the file, and Ctrl+X y enter to save
user@host:~/$ cat db_password
1234
user@host:~/$ # this implies a new line was added
user@host:~/$ rm db_password
user@host:~/$ nano -L db_password
# enter "1234" into the file, and Ctrl+X y enter to save
user@host:~/$ cat db_password
1234user@host:~/$ # no new line was added to the end of the file
```

The true issue @Joffcom is that n8n doesn't ignore the newline character at the end of a docker secrets file, particularly if edited with `nano`. It would appear other applications that I use do, in fact, do this, because I never use `nano` with `-L` flag and haven't had any issues with other applications.

Editing the password secrets file with `nano -L` and deleting the newline after the password allowed the n8n container to start as well as postgresql reporting proper authentication. I imagine it's the same issue with other docker secrets, and I'm off to start debugging those with `nano -L`

**Comment by elcid2012** [comment]
Yes, in fact `DB_POSTGRESDB_USER_FILE` also has the same issue when editing with `nano` without the `-L` flag, because the postgresql container, which is already initialized with the username "n8n" throws this error:

```
2025-08-03 14:18:00.342 EDT [922] DETAIL:  Role "n8n
	" does not exist.
	Connection matched file "/var/lib/postgresql/data/pg_hba.conf" line 128: "host all all all scram-sha-256"
```

where the `DB_POSTGRESDB_USER_FILE` contains simply `n8n`

The error log indicates that the newline within the docker secret is being read by n8n and passed as the username, newline and all, to postgresql.

**Comment by fredericopadrao** [comment]
> We're all circling around the answer here, but using the clues I searched and found this: https://askubuntu.com/questions/13317/how-to-stop-gedit-gvim-vim-nano-from-adding-end-of-file-newline-char
> 
> Here's the actual fix:
> 
> `nano -L db_password`
> 
> Then insert your password into the file, and save as usual.
> 
> The `-L` flag forces nano to NOT insert new lines at the end of the file. You can verify this with this psuedo code:
> 
> ```
> user@host:~/$ nano db_password
> # enter "1234" into the file, and Ctrl+X y enter to save
> user@host:~/$ cat db_password
> 1234
> user@host:~/$ # this implies a new line was added
> user@host:~/$ rm db_password
> user@host:~/$ nano -L db_password
> # enter "1234" into the file, and Ctrl+X y enter to save
> user@host:~/$ cat db_password
> 1234user@host:~/$ # no new line was added to the end of the file
> ```
> 
> The true issue [@Joffcom](https://github.com/Joffcom) is that n8n doesn't ignore the newline character at the end of a docker secrets file, particularly if edited with `nano`. It would appear other applications that I use do, in fact, do this, because I never use `nano` with `-L` flag and haven't had any issues with other applications.
> 
> Editing the password secrets file with `nano -L` and deleting the newline after the password allowed the n8n container to start as well as postgresql reporting proper authentication. I imagine it's the same issue with other docker secrets, and I'm off to start debugging those with `nano -L`

Yeap, that solved the problem, nano -L. All perfect over here! Thanksss!
